### PR TITLE
Prioritize shorter segments over longer ones, if equal priority

### PR DESCRIPTION
**Pull request by TheBB**

### CODE REVIEW

**Comment by TheBB** [comment]
This is probably going to mess things up because the order might change, no longer any guarantee that hidden and shown segments are ordered together in the list any more.

**Comment by TheBB** [comment]
Updated. @deb0ch thoughts?

**Comment by dsedivec** [comment]
FWIW I've been running this PR's code for a while, and unless I'm mistaken, cranking the `(buffer-id ...)` set in `spaceline-spacemacs-theme` up to `:priority 10` makes it always display.  Setting `:priority -1` or -2 or -3 or -10 has no effect.  Not sure if -1 is still supposed to be the lowest priority, but in any case I was surprised that a *higher* priority seemed to make `buffer-id` and friends always part of the mode line, rather than a *lower* priority.

It is, of course, very possible that I'm doing something wrong.

**Comment by deb0ch** [comment]
Hey there, sorry for the huge delay but I was caught up in another, very intense project.

Unless I'm mistaken this PR does several things at once:

- change the sorting function to take segment length into account (line 668 to 674)
- make the sorting function evaluated every time the spaceline is rendered, instead of only once at compile time (line 644)
- somehow change the hiding / showing code. This is the part that I don't really get (line 648 to 659)
- give the buffer-id segment the default priority

I think that to achieve what the title says only the first two points are necessary.

I think that the second point implies a performance penalty, however I don't know how big.

The last point why not, I actually miss it very much when I don't see the title of the file I'm editing.

I'm left very confused with the third point. I do not understand how this change affects the functionality of this code. I also find that this version is harder to read, and if anything will make us loose a little bit of performance since all the list is always scanned instead of stopping when having a satisfying length. If the segments are correctly sorted it is not needed to go through the entire list.

I also think that taking into account the length of the segment for sorting takes away a bit of the simplicity of this mechanism, making it harder for the user to predict what should appear or disappear.

On a last note, I think that the underlying issue, being the buffer-id taking up too much space if the filename is too long, should be solved differently. It would make more sense, IMHO, to limit the displayed length of the buffer-id in the modeline, for instance by trimming the middle characters.

For instance, `much-much-to-long-buffer-id-03` would become `much-m...-id-03`.

**TL;DR:** 
- code and behavior are more complex,
- we loose on performance,
- there's a part that left me confused,
- I think the underlying issue should be solved differently.

**Comment by TheBB** [comment]
Closing this. Thanks @deb0ch
